Transform Gizmos on Mirabuf Assemblies

Dhruv-0-Arora opened the pull request:
### Description
This branch adds transform gizmos to the mirabuf assemblies as they spawn in. These transform gizmos currently are always on the Mirabuf assemblies as they spawn. The physics system is disabled when the user presses the transform gizmo (will implement so physics is disabled for only the one assembly and also it stays disable as long as the gizmo is attached).

[JIRA Issue](https://jira.autodesk.com/browse/AARD-1733)

Dhruv-0-Arora committed:
Adding disable physics and enable physics functionality

Dhruv-0-Arora committed:
Adding mesh to the sceneobject

Dhruv-0-Arora committed:
Disabling Physics system when using transform gizmo

Dhruv-0-Arora committed:
Optimized the enabling and disabling of the physics
Made the functions only run once by putting them in the event listener

Dhruv-0-Arora committed:
Finished transform gizmo

Dhruv-0-Arora commented:
Right now it does not set Rotation or Scale. Only position

Dhruv-0-Arora committed:
Ran Prettier and fixed all issues with having multiple gizmos on the same object and also implemented physics being disabled only for the one object
need to implement esc listener to escape the transform gizmo

Dhruv-0-Arora committed:
Added removing gizmo functionality in MainHUD

Dhruv-0-Arora committed:
Added implementation of enter and escape key and merged remote-tracking branch 'origin/dev' into dhruv/1733/mirabuf-gizmo
made a lot of changes with bugs and other issues

Dhruv-0-Arora committed:
Ran prettier

Dhruv-0-Arora committed:
Merge remote-tracking branch 'origin/dev' into dhruv/1733/mirabuf-gizmo

Dhruv-0-Arora committed:
Moved all transform gizmo implementation into TransformGizmos file for handling

Dhruv-0-Arora committed:
Moved shift listener functionality to InputSystem for optimization

Dhruv-0-Arora commented:
Completely ready for review

HunterBarclay committed:
Fixed collision contacts for when moving the robot

HunterBarclay reviewed (changes requested):
I added a quick change to your PR that also sets the bodies to be sensors when disabled, meaning they won't act on generated collision contacts.

When spawning in a field, we should restrict it to translation movement only. Not rotation. I also noticed that after finishing with the gizmo, if I try to move the camera by clicking near the center of the screen, it doesn't work, I have to click on the edge, and then it essentially snaps out of it and I can orbit the camera like normal.

HunterBarclay committed:
Updated workflow

Dhruv-0-Arora committed:
Removed rotation gizmo from mirabufsceneobject

Dhruv-0-Arora committed:
Merge remote-tracking branch 'refs/remotes/origin/dhruv/1733/mirabuf-gizmo' into dhruv/1733/mirabuf-gizmo

Dhruv-0-Arora committed:
Fixed orbit bug

Dhruv-0-Arora committed:
Ran Prettier

HunterBarclay commented on fission/src/mirabuf/MirabufSceneObject.ts:
```diff
@@ -77,11 +99,36 @@ class MirabufSceneObject extends SceneObject {
+                }
+
+                // disable physics for the body
+                World.PhysicsSystem.DisablePhysicsForBody(this._mechanism.GetBodyByNodeId(rn.id)!)
```
Having this run every frame probably isn't the best idea.

Dhruv-0-Arora replied:
Oh what I forgot to change that

HunterBarclay commented on fission/src/mirabuf/MirabufSceneObject.ts:
```diff
@@ -77,11 +99,36 @@ class MirabufSceneObject extends SceneObject {
+                // commands to either cancel gizmo creation or confirm position
+                if (InputSystem.getInput("enter")) {
+                    this.transformGizmos.removeGizmos()
+                    World.PhysicsSystem.EnablePhysicsForBody(this._mechanism.GetBodyByNodeId(this._mechanism.rootBody)!)
```
```suggestion
                    World.PhysicsSystem.EnablePhysicsForBody(this._mechanism.GetBodyByNodeId(rn.id)!)
```

HunterBarclay reviewed (changes requested):
Nearly there, just a few tweaks. After positioning everything, the robot doesn't interact with the gampieces and the arm of mean machine doesn't interact with anything. Fix is added.

Dhruv-0-Arora committed:
Fixed enabling and disabling physics iteratively

Dhruv-0-Arora committed:
Minor changes for transform gizmo and fixing potential future bugs
Added deattaching transform gizmos from mesh

Dhruv-0-Arora committed:
Ran prettier

HunterBarclay commented on fission/src/mirabuf/MirabufSceneObject.ts:
```diff
@@ -77,11 +106,35 @@ class MirabufSceneObject extends SceneObject {
+            /**
+             * transform gizmo block to update the position and rotation of the body to match the gizmos's position
+             */
+            if (this.transformGizmos.isActive) {
+                // commands to either cancel gizmo creation or confirm position
+                if (InputSystem.getInput("enter")) {
+                    this.transformGizmos.RemoveGizmos()
+                    World.PhysicsSystem.EnablePhysicsForBody(this._mechanism.GetBodyByNodeId(rn.id)!)
+                    World.PhysicsSystem.EnablePhysicsForBody(this._mechanism.GetBodyByNodeId(this._mechanism.rootBody)!)
+                    return
+                } else if (InputSystem.getInput("escape")) {
+                    this.transformGizmos.RemoveGizmos()
+                    World.SceneRenderer.RemoveSceneObject(this.id)
+                    return
+                }
+
+                // if the gizmo is being dragged, copy the mesh position and rotation to the Mirabuf body
+                if (this.transformGizmos.isBeingDragged()) {
+                    this.transformGizmos.UpdateMirabufPositioning(this, rn)
+                    World.PhysicsSystem.DisablePhysicsForBody(this._mechanism.GetBodyByNodeId(rn.id)!)
+                }
+            }
```
You should move this to only be done once per update, not for each node. You're running into an issue where enter and escape are only being detected in one of these loops before being discarded so running enable on all the rigid nodes doesn't work.

HunterBarclay commented on fission/src/mirabuf/MirabufSceneObject.ts:
```diff
@@ -37,6 +49,22 @@ class MirabufSceneObject extends SceneObject {
+        this.transformGizmos.CreateGizmo("translate")
+
+        // disabling physics initially for all rigid node bodies
+        this._mirabufInstance.parser.rigidNodes.forEach(rn => {
```
Please turn adding and removing the gizmos and essentially "putting the mirabuf object in move mode" into a separate function for reusability

Dhruv-0-Arora committed:
Merge remote-tracking branch 'origin/dev' into dhruv/1733/mirabuf-gizmo

Dhruv-0-Arora committed:
Ran prettier

Dhruv-0-Arora committed:
Removed testing code

Dhruv-0-Arora committed:
Adding bug fixes from JIRA 1728

Dhruv-0-Arora committed:
Quick fix for Enabling Physics
enables physics for all nodes
also removed redudant enabling physics for root node

Dhruv-0-Arora committed:
Merge remote-tracking branch 'origin/dev' into dhruv/1733/mirabuf-gizmo

Dhruv-0-Arora committed:
Run prettier

PepperLola reviewed (commented):
Are the negative axis arrows supposed to be partially invisible? Also, I noticed the squares for moving along a plane disappear when the angle between them and the camera is too shallow. Not sure if there's much you can do about that though but I could be wrong.

Dhruv-0-Arora committed:
Merge remote-tracking branch 'origin/dev' into dhruv/1733/mirabuf-gizmo

Dhruv-0-Arora committed:
Moved transform gizmo spawning into public function

Dhruv-0-Arora commented:
> Are the negative axis arrows supposed to be partially invisible? Also, I noticed the squares for moving along a plane disappear when the angle between them and the camera is too shallow. Not sure if there's much you can do about that though but I could be wrong.

Well I can't really do much about either since it is the built in transform gizmo from three.js. The negative axis arrows being partially invisible is intentional on their end

Dhruv-0-Arora committed:
Removed functionality of 'esc' after first placing gizmo

Dhruv-0-Arora committed:
Added disable physics function and make optimization change in SceneRenderer

Dhruv-0-Arora committed:
Ran prettier

HunterBarclay commented on fission/src/Window.d.ts:
This file was removed in dev, what's going on with this file?

HunterBarclay replied:
@PepperLola You know what's up with this file?

HunterBarclay commented on fission/src/mirabuf/MirabufSceneObject.ts:
```diff
@@ -27,6 +29,17 @@ class MirabufSceneObject extends SceneObject {
     private _physicsLayerReserve: LayerReserve | undefined = undefined
 
+    private _transformGizmos: TransformGizmos | undefined = undefined
+    private _deleteGizmoOnEscape: boolean = true
```
I'll let this be for now, but the entirety of transform gizmos needs a major refactor to be more adaptable. It should honestly be worked into some kind of UI component. [Jira ticket for refactoring](https://jira.autodesk.com/browse/AARD-1744)

HunterBarclay commented on fission/src/ui/components/TransformGizmos.tsx:
This isn't a UI component as it stands atm.

HunterBarclay reviewed (approved):
I'm approving this for now because we want it in for the MVP and it's currently blocking two critical tasks. That being said I created a new Jira ticket for Sprint 4 to address a lot of the issues I have with Transform Gizmos.

Dhruv-0-Arora committed:
Removed window.d.ts again

BrandonPacewic reviewed (approved):
Let's get this MVP done!
